Add Prometheus metrics and profiling information.

NullHypothesis opened the pull request:
This PR adds two new command line flags. One exposes Prometheus metrics and the other enables profiling. The git commit messages explain this work in detail.

rillian commented on enclave.go:
```diff
@@ -269,14 +306,21 @@ func (e *Enclave) Stop() error {
-	elog.Printf("Starting public (%s) and private (%s) Web server.", e.pubSrv.Addr, e.privSrv.Addr)
+	if e.cfg.PrometheusPort > 0 {
+		elog.Printf("Starting Prometheus Web server (%s).", e.promSrv.Addr)
+		go e.promSrv.ListenAndServe() //nolint:errcheck
```
This should be `ListenAndServeTLS` I think, since it's facing outside the enclave. Even if we don't publicly  route the port, we should encrypt transfer between machines.

NullHypothesis replied:
Addressed in feaf133483114ec649785bc4e3b6d41173f963f6.

rillian reviewed (approved):
In what way are the Prometheus metrics coarse-grained? I thought it just returned whatever the current counter values were.

r=me with comments addressed.

NullHypothesis commented on enclave.go:
```diff
@@ -168,6 +170,15 @@ func (c *Config) Validate() error {
 
+// String returns a string representation of the enclave's configuration.
+func (c *Config) String() string {
+	s, err := json.MarshalIndent(c, "", "\t")
```
Addressed in f0ce959bcafa7d5d27d023c78fa639e5901b8feb.

NullHypothesis commented:
@rillian: I addressed your comments and also made Prometheus capture requests for the internal Web server. The Prometheus and the public Web server now share a TLS config. This means that the Prometheus server is going to be unavailable until the public Web server obtained an HTTPS certificate, but that should be fine.

NullHypothesis commented:
I ended up refactoring the PR in various ways:

- Make nitriding fail if any of our Web servers fail to start (64d4624e00dd127c74730da678eb2c453ba15883). I noticed this issue by accident, when I tried to start two Web servers on the same port, and nitriding failed silently.
- Register Prometheus's default "Go" and "process" collector. These are going to give us insight into process load (6ca66fb97c0f7b5b37195b294c88f8561ca89074).
- Split up our custom metrics into "HTTP requests targeting nitriding itself" and "HTTP requests targeting the enclave application" (4ec9d8d675e3cb61da49033f22412cf0304821ab).

rillian commented:
@NullHypothesis Sorry if I missed it, but did you answer my question about how the metrics coarse-graining works?

NullHypothesis commented:
> @NullHypothesis Sorry if I missed it, but did you answer my question about how the metrics coarse-graining works?

They are coarse-grained in the sense that we only get to learn about HTTP request responses (these are the ones we instrument ourselves) plus some runtime information which is provided by Prometheus itself. Still, this probably warrants a privacy review. (cc @ShivanKaul)

ShivanKaul commented:
Could you clarify what data we'll be seeing?

NullHypothesis commented:
> Could you clarify what data we'll be seeing?

I intend to expose two types of information:

1. HTTP requests to nitriding and star-randsrv
    - Capture the method, path, response body (iff it's an HTTP error), and response code of each request, e.g., we will be able to tell that there have been 50 GET requests to /info that resulted in a 200 status code; or 2 PUT requests to /state that resulted in a 500 status code.
3. Enclave-internal runtime information, e.g.:
    - Number of currently-active goroutines
    - Virtual and resident memory size
    - User/system CPU time in seconds
    - Number of open file descriptors
    - Garbage collector statistics

We could do without HTTP requests but some sort of runtime information is important because it helps us learn when the enclave is reaching its computational limits. In the context of star-randsrv, both types of data strike me as fine because it doesn't leak any information about the randomness server's secret key. Can we think of a case where the above data would jeopardize the randomness server's security or privacy guarantees?

ShivanKaul commented:
This seems fairly innocuous, I can't think of how this could be used adversarially without a lot of malicious effort.

rillian commented on cmd/main.go:
```diff
@@ -56,15 +60,20 @@ func main() {
+		IntPort:        uint16(intPort),
+		PrometheusPort: uint16(prometheusPort),
+		HostProxyPort:  uint32(hostProxyPort),
+		UseProfiling:   useProfiling,
+		UseACME:        useACME,
+		WaitForApp:     waitForApp,
+		Debug:          debug,
```
Can `UseProfiling` be just above `Debug` to keep related values together?

rillian commented on enclave.go:
```diff
@@ -330,19 +330,32 @@ func (e *Enclave) Stop() error {
 func startWebServers(e *Enclave) error {
 	if e.cfg.PrometheusPort > 0 {
 		elog.Printf("Starting Prometheus Web server (%s).", e.promSrv.Addr)
-		go e.promSrv.ListenAndServeTLS("", "") //nolint:errcheck
```
So `nolint` was a bug?

NullHypothesis replied:
It was meant to silence the linter, which complains about unchecked error returns. Since I now decided that we should check these returns, it's fair to call this a bug.

rillian reviewed (dismissed):
Ok, thanks for clarifying. Seems reasonable to me.

Thanks for the additional review, @ShivanKaul!

rillian reviewed (approved):
looks good, thanks!
